File: multipicker/src/main/java/com/edwardstock/multipicker/internal/MediaFileLoader.java
package com.edwardstock.multipicker.internal;

import android.content.Context;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.media.MediaMetadataRetriever;
import android.provider.MediaStore;
import android.support.annotation.Nullable;

import com.annimon.stream.Stream;
import com.edwardstock.multipicker.PickerConfig;
import com.edwardstock.multipicker.data.Dir;
import com.edwardstock.multipicker.data.MediaFile;
import com.edwardstock.multipicker.data.MediaSize;
import com.edwardstock.multipicker.data.VideoInfo;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import timber.log.Timber;


public class MediaFileLoader {

    private final String[] mPhotoProjection = new String[]{
            MediaStore.Images.Media._ID,
            MediaStore.Images.Media.DISPLAY_NAME,
            MediaStore.Images.Media.DATA,
            MediaStore.Images.Media.BUCKET_DISPLAY_NAME,
            MediaStore.Images.Media.WIDTH,
            MediaStore.Images.Media.HEIGHT,
    };
    private final String[] mPhotoAndVideoProjection = new String[]{
            MediaStore.Images.Media._ID,
            MediaStore.Images.Media.DISPLAY_NAME,
            MediaStore.Images.Media.DATA,
            MediaStore.Images.Media.BUCKET_DISPLAY_NAME,
            MediaStore.Video.Media.WIDTH,
            MediaStore.Video.Media.HEIGHT,
            MediaStore.Video.Media.DURATION,
            MediaStore.Images.Media.WIDTH,
            MediaStore.Images.Media.HEIGHT,
    };
    private final String[] mVideoProjection = new String[]{
            MediaStore.Video.Media._ID,
            MediaStore.Video.Media.DISPLAY_NAME,
            MediaStore.Video.Media.DATA,
            MediaStore.Video.Media.BUCKET_DISPLAY_NAME,
            MediaStore.Video.Media.WIDTH,
            MediaStore.Video.Media.HEIGHT,
            MediaStore.Video.Media.DURATION,
    };
    private Context mContext;
    private ExecutorService mExecutorService;

    public MediaFileLoader(Context context) {
        mContext = context;
    }

    @Nullable
    private static File makeSafeFile(String path) {
        if (path == null || path.isEmpty()) {
            return null;
        }
        try {
            return new File(path);
        } catch (Exception ignored) {
            return null;
        }
    }

    public void loadDeviceImages(PickerConfig config, final OnLoadListener listener) {
        getExecutorService().execute(new ImageLoadRunnable(config, listener));
    }

    public void abortLoadImages() {
        if (mExecutorService != null) {
            mExecutorService.shutdown();
            mExecutorService = null;
        }
    }

    private ExecutorService getExecutorService() {
        if (mExecutorService == null) {
            mExecutorService = Executors.newSingleThreadExecutor();
        }
        return mExecutorService;
    }

    private Bitmap createVideoThumbnail(String filePath) {
        MediaMetadataRetriever mmr = new MediaMetadataRetriever();
        mmr.setDataSource(filePath);
        //api time unit is microseconds
        try {
            return mmr.getFrameAtTime(0, MediaMetadataRetriever.OPTION_CLOSEST_SYNC);
        } finally {
            mmr.release();
        }
    }

    public interface OnLoadListener {
        void onFilesLoadFailed(Throwable t);
        void onFilesLoadedSuccess(List<MediaFile> images, List<Dir> dirList);
    }

    private class ImageLoadRunnable implements Runnable {
        private PickerConfig mConfig;
        private OnLoadListener mImageLoadListener;
        private String mRoot = "external";

        public ImageLoadRunnable(final PickerConfig config, OnLoadListener imageLoadListener) {
            mConfig = config;
            mImageLoadListener = imageLoadListener;
        }

        @Override
        public void run() {
            Cursor cursor;
            final String[] projection;
            String selection;

            if (mConfig.isShowVideos() && mConfig.isShowPhotos()) {
                projection = mPhotoAndVideoProjection;
                selection = MediaStore.Files.FileColumns.MEDIA_TYPE + "=" + MediaStore.Files.FileColumns.MEDIA_TYPE_IMAGE +
                        " OR "
                        + MediaStore.Files.FileColumns.MEDIA_TYPE + "=" + MediaStore.Files.FileColumns.MEDIA_TYPE_VIDEO;

                cursor = mContext.getContentResolver().query(MediaStore.Files.getContentUri(mRoot), projection,
                        selection, null, MediaStore.Files.FileColumns.DATE_ADDED);
            } else if (!mConfig.isShowPhotos() && mConfig.isShowVideos()) {
                projection = mVideoProjection;
                cursor = mContext.getContentResolver().query(
                        MediaStore.Video.Media.EXTERNAL_CONTENT_URI, projection,
                        null, null, MediaStore.Files.FileColumns.DATE_ADDED);
            } else {
                projection = mPhotoProjection;
                cursor = mContext.getContentResolver().query(
                        MediaStore.Images.Media.EXTERNAL_CONTENT_URI, projection,
                        null, null, MediaStore.Files.FileColumns.DATE_ADDED);
            }


            if (cursor == null) {
                mImageLoadListener.onFilesLoadFailed(new NullPointerException());
                return;
            }

            List<MediaFile> files = new ArrayList<>();
            Map<String, Dir> dirMap = new HashMap<>();

            int cntVideos = 0;

            if (cursor.moveToLast()) {
                do {
                    long id = cursor.getLong(cursor.getColumnIndex(projection[0]));
                    String name = cursor.getString(cursor.getColumnIndex(projection[1]));
                    String path = cursor.getString(cursor.getColumnIndex(projection[2]));
                    String bucket = cursor.getString(cursor.getColumnIndex(projection[3]));
                    String videoWidth = null;
                    String videoHeight = null;
                    String videoDuration = null;
                    String imageWidth = null;
                    String imageHeight = null;

                    if (mConfig.isShowPhotos() && mConfig.isShowVideos()) {
                        videoWidth = cursor.getString(cursor.getColumnIndex(projection[4]));
                        videoHeight = cursor.getString(cursor.getColumnIndex(projection[5]));
                        videoDuration = cursor.getString(cursor.getColumnIndex(projection[6]));
                        imageWidth = cursor.getString(cursor.getColumnIndex(projection[7]));
                        imageHeight = cursor.getString(cursor.getColumnIndex(projection[8]));
                    } else if (!mConfig.isShowPhotos() && mConfig.isShowVideos()) {
                        videoWidth = cursor.getString(cursor.getColumnIndex(projection[4]));
                        videoHeight = cursor.getString(cursor.getColumnIndex(projection[5]));
                        videoDuration = cursor.getString(cursor.getColumnIndex(projection[6]));
                    } else {
                        imageWidth = cursor.getString(cursor.getColumnIndex(projection[4]));
                        imageHeight = cursor.getString(cursor.getColumnIndex(projection[5]));
                    }

                    File file = makeSafeFile(path);
                    if (file == null) {
                        continue;
                    }

                    if (mConfig.getExcludedFiles().contains(file.getAbsolutePath())) {
                        continue;
                    }

                    if (file.isFile()) {
                        MediaFile mediaFile;
                        if (PickerUtils.isVideoFormat(path)) {
                            cntVideos++;
                            mediaFile = new MediaFile(id, name, path,
                                    new VideoInfo(new MediaSize(videoWidth, videoHeight), videoDuration)
                            );
                        } else {
                            mediaFile = new MediaFile(id, name, path, new MediaSize(imageWidth, imageHeight));
                        }

                        files.add(mediaFile);

                        Dir folder = dirMap.get(bucket);
                        if (folder == null) {
                            folder = new Dir(bucket);
                            dirMap.put(bucket, folder);
                        }
                        folder.getFiles().add(mediaFile);
                    }

                } while (cursor.moveToPrevious());
            }
            cursor.close();

            if (cntVideos > 0) {
                getExecutorService().execute(new VideoThumbResolver(files));
            }

            /* Convert HashMap to ArrayList if not null */
            List<Dir> dirs = new ArrayList<>(dirMap.values());

            mImageLoadListener.onFilesLoadedSuccess(files, dirs);
        }
    }

    private final class VideoThumbResolver implements Runnable {
        private final List<MediaFile> mFiles;

        VideoThumbResolver(List<MediaFile> files) {
            mFiles = Stream.of(files).withoutNulls().filter(MediaFile::isVideo).toList();
        }

        @Override
        public void run() {
            for (final MediaFile item : mFiles) {
                final File f = PickerUtils.createVideoThumbFile(item);
                if (f == null) {
                    return;
                }

                final Bitmap bmp = createVideoThumbnail(item.getPath());
                try (FileOutputStream o = new FileOutputStream(f)) {
                    Timber.d("Write thumbnail for video %s: %s", item.getPath(), f.getAbsolutePath());
                    bmp.compress(Bitmap.CompressFormat.JPEG, 100, o);
                } catch (FileNotFoundException e) {
                    e.printStackTrace();
                } catch (IOException e) {
                    e.printStackTrace();
                } finally {
                    bmp.recycle();
                }
            }
        }
    }

}

File: multipicker/src/main/java/com/edwardstock/multipicker/internal/MediaFileLoader.kt
package com.edwardstock.multipicker.internal

import android.content.ContentUris
import android.content.Context
import android.database.Cursor
import android.graphics.Bitmap
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.os.Build
import android.provider.MediaStore
import com.edwardstock.multipicker.PickerConfig
import com.edwardstock.multipicker.data.Dir
import com.edwardstock.multipicker.data.MediaFile
import com.edwardstock.multipicker.data.MediaSize
import com.edwardstock.multipicker.data.VideoInfo
import timber.log.Timber
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class MediaFileLoader(private val ctx: Context) {
    private val photoProjection = arrayOf(
            MediaStore.Images.Media._ID,
            MediaStore.Images.Media.DISPLAY_NAME,
            MediaStore.Images.Media.DATA,
            MediaStore.Images.Media.BUCKET_DISPLAY_NAME,
            MediaStore.Images.Media.WIDTH,
            MediaStore.Images.Media.HEIGHT)
    private val photoAndVideoProjection = arrayOf(
            MediaStore.Images.Media._ID,
            MediaStore.Images.Media.DISPLAY_NAME,
            MediaStore.Images.Media.DATA,
            MediaStore.Images.Media.BUCKET_DISPLAY_NAME,
            MediaStore.Video.Media.WIDTH,
            MediaStore.Video.Media.HEIGHT,
            MediaStore.Video.Media.DURATION,
            MediaStore.Images.Media.WIDTH,
            MediaStore.Images.Media.HEIGHT)

    private val videoProjection = arrayOf(
            MediaStore.Video.Media._ID,
            MediaStore.Video.Media.DISPLAY_NAME,
            MediaStore.Video.Media.DATA,
            MediaStore.Video.Media.BUCKET_DISPLAY_NAME,
            MediaStore.Video.Media.WIDTH,
            MediaStore.Video.Media.HEIGHT,
            MediaStore.Video.Media.DURATION)


    private val executorService: ExecutorService by lazy {
        Executors.newSingleThreadExecutor()
    }

    fun loadDeviceImages(config: PickerConfig, listener: OnLoadListener) {
        executorService.execute(ImageLoadRunnable(config, listener))
    }

    fun abortLoadImages() {
        executorService.shutdown()
    }

    private fun createVideoThumbnail(filePath: String): Bitmap? {
        val mmr = MediaMetadataRetriever()
        //api time unit is microseconds
        return try {
            mmr.setDataSource(filePath)
            mmr.getFrameAtTime(0, MediaMetadataRetriever.OPTION_CLOSEST_SYNC)
        } catch (e: Throwable) {
            Timber.e(e, "Failed to create video thumbnail")
            null
        } finally {
            mmr.release()
        }
    }

    interface OnLoadListener {
        fun onFilesLoadFailed(t: Throwable?)
        fun onFilesLoadedSuccess(images: List<MediaFile>, dirList: List<Dir>)
    }

    private inner class ImageLoadRunnable(val config: PickerConfig, val imageLoadListener: OnLoadListener) : Runnable {
        private val root = "external"

        private fun isOsGteQ(): Boolean {
            return Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q
        }

        override fun run() {
            val cursor: Cursor?
            val projection: Array<String>
            val selection: String

            if (config.isShowVideos && config.isShowPhotos) {
                projection = photoAndVideoProjection
                // select media_type=1 or media_type=3
                selection = (MediaStore.Files.FileColumns.MEDIA_TYPE + "=" + MediaStore.Files.FileColumns.MEDIA_TYPE_IMAGE +
                        " OR "
                        + MediaStore.Files.FileColumns.MEDIA_TYPE + "=" + MediaStore.Files.FileColumns.MEDIA_TYPE_VIDEO)

                cursor = ctx.contentResolver.query(
                        MediaStore.Files.getContentUri(root),
                        projection,
                        selection,
                        null,
                        MediaStore.Files.FileColumns.DATE_ADDED
                )
            } else if (!config.isShowPhotos && config.isShowVideos) {
                projection = videoProjection
                cursor = ctx.contentResolver.query(
                        MediaStore.Video.Media.EXTERNAL_CONTENT_URI,
                        projection,
                        null,
                        null,
                        MediaStore.Files.FileColumns.DATE_ADDED
                )
            } else {
                projection = photoProjection
                cursor = ctx.contentResolver.query(
                        MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                        projection,
                        null,
                        null,
                        MediaStore.Files.FileColumns.DATE_ADDED
                )
            }
            if (cursor == null) {
                imageLoadListener.onFilesLoadFailed(NullPointerException())
                return
            }
            val files: MutableList<MediaFile> = ArrayList<MediaFile>()
            val dirMap: MutableMap<String, Dir> = HashMap<String, Dir>()
            var cntVideos = 0
            if (cursor.moveToLast()) {
                do {
                    val id = cursor.getLong(cursor.getColumnIndexOrThrow(projection[0]))
                    val name = cursor.getString(cursor.getColumnIndexOrThrow(projection[1]))
                    val path = cursor.getString(cursor.getColumnIndexOrThrow(projection[2]))
                    val bucket = cursor.getString(cursor.getColumnIndexOrThrow(projection[3]))
                    var videoWidth: String? = null
                    var videoHeight: String? = null
                    var videoDuration: String? = null
                    var imageWidth: String? = null
                    var imageHeight: String? = null
                    if (config.isShowPhotos && config.isShowVideos) {
                        videoWidth = cursor.getString(cursor.getColumnIndexOrThrow(projection[4]))
                        videoHeight = cursor.getString(cursor.getColumnIndexOrThrow(projection[5]))
                        videoDuration = cursor.getString(cursor.getColumnIndexOrThrow(projection[6]))
                        imageWidth = cursor.getString(cursor.getColumnIndexOrThrow(projection[7]))
                        imageHeight = cursor.getString(cursor.getColumnIndexOrThrow(projection[8]))
                    } else if (!config.isShowPhotos && config.isShowVideos) {
                        videoWidth = cursor.getString(cursor.getColumnIndexOrThrow(projection[4]))
                        videoHeight = cursor.getString(cursor.getColumnIndexOrThrow(projection[5]))
                        videoDuration = cursor.getString(cursor.getColumnIndexOrThrow(projection[6]))
                    } else {
                        imageWidth = cursor.getString(cursor.getColumnIndexOrThrow(projection[4]))
                        imageHeight = cursor.getString(cursor.getColumnIndexOrThrow(projection[5]))
                    }
                    val file = makeSafeFile(path) ?: continue
                    if (config.excludedFiles.contains(file.absolutePath)) {
                        continue
                    }
                    if (file.isFile) {

                        val mediaFile: MediaFile = if (PickerUtils.isVideoFormat(path)) {
                            cntVideos++
                            val contentUri: Uri = ContentUris.withAppendedId(
                                    MediaStore.Video.Media.EXTERNAL_CONTENT_URI,
                                    id
                            )
                            MediaFile(id, name, path, contentUri, videoInfo = VideoInfo(MediaSize(videoWidth, videoHeight), videoDuration))
                        } else {
                            val contentUri: Uri = ContentUris.withAppendedId(
                                    MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                                    id
                            )
                            MediaFile(id, name, path, contentUri, mediaSize = MediaSize(imageWidth, imageHeight))
                        }
                        files += mediaFile
                        var folder: Dir? = dirMap[bucket]
                        if (folder == null) {
                            if (bucket != null) {
                                folder = Dir(bucket)
                                dirMap[bucket] = folder
                            }
                        }
                        folder?.let {
                            it.files += mediaFile
                        }
                    }
                } while (cursor.moveToPrevious())
            }
            cursor.close()
            if (cntVideos > 0) {
                executorService.execute(VideoThumbResolver(files))
            }

            /* Convert HashMap to ArrayList if not null */
            val dirs: List<Dir> = ArrayList(dirMap.values)
            imageLoadListener.onFilesLoadedSuccess(files, dirs)
        }
    }

    private inner class VideoThumbResolver(files: List<MediaFile>) : Runnable {
        var files = files.filter { it.isVideo }

        override fun run() {
            for (item in files) {
                val thumbFile = PickerUtils.createVideoThumbFile(ctx, item) ?: return

                var bmp: Bitmap? = null
                try {
                    bmp = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                        ctx.contentResolver.loadThumbnail(item.uri, android.util.Size(640, 640), null)
                    } else {
                        createVideoThumbnail(item.uri) ?: continue
                    }

                    FileOutputStream(thumbFile).use { o ->
                        Timber.d("Write thumbnail for video %s: %s", item.uri, thumbFile.absolutePath)
                        bmp.compress(Bitmap.CompressFormat.JPEG, 100, o)
                    }
                } catch (e: FileNotFoundException) {
                    Timber.e(e, "Unable to create or save thumbnail")
                } catch (e: IOException) {
                    Timber.e(e, "Unable to create or save thumbnail")
                } finally {
                    bmp?.recycle()
                }
            }
        }
    }

    companion object {
        private fun makeSafeFile(path: String?): File? {
            return if (path == null || path.isEmpty()) {
                null
            } else try {
                File(path)
            } catch (ignored: Exception) {
                null
            }
        }
    }
}